File: android/src/main/kotlin/com/netki/netki_sdk/Mappers.kt
package com.netki.netki_sdk

import com.netki.model.AdditionalDataField
import com.netki.model.BusinessConfiguration
import com.netki.model.IdCountry
import com.netki.model.IdType
import com.netki.model.OnBoardIdTheme
import com.netki.model.Picture
import com.netki.model.PictureType
import com.netki.model.RequestStatus
import com.netki.model.ResultInfo

object Mappers {

    // MARK: - ResultInfo

    fun resultInfoToMap(resultInfo: ResultInfo): Map<String, Any?> {
        return mapOf(
            "status" to if (resultInfo.status == RequestStatus.SUCCESS) "SUCCESS" else "ERROR",
            "extraData" to sanitizeExtraData(resultInfo.extraData),
            "errorType" to resultInfo.errorType?.name,
            "message" to resultInfo.message
        )
    }

    // Recursively sanitize extraData to ensure all values are Flutter-compatible
    private fun sanitizeExtraData(data: Any?): Any? {
        return when (data) {
            null -> null
            is String, is Number, is Boolean -> data
            is Map<*, *> -> data.mapValues { sanitizeExtraData(it.value) }
            is List<*> -> data.map { sanitizeExtraData(it) }
            is com.netki.model.Picture -> pictureToMap(data)
            is com.netki.model.Barcode -> barcodeToMap(data)
            is com.netki.model.Barcode.DriverLicense -> driverLicenseToMap(data)
            is com.netki.model.PassportContent -> passportContentToMap(data)
            is com.netki.model.EPassportContent -> ePassportContentToMap(data)
            is com.netki.model.AdditionalPersonalData -> additionalPersonalDataToMap(data)
            is com.netki.model.LivenessInformation -> livenessInformationToMap(data)
            is Enum<*> -> data.name
            else -> data.toString() // Fallback: convert to string
        }
    }

    // MARK: - IdCountry

    fun idCountryToMap(country: IdCountry): Map<String, Any?> {
        return mapOf(
            "name" to country.name,
            "alpha2" to country.alpha2,
            "alpha3" to country.alpha3,
            "countryCallingCode" to country.countryCallingCode,
            "hasBarcodeId" to country.hasBarcodeId,
            "flag" to country.flag,
            "isBanned" to country.isBanned,
            "hasNfcPassport" to country.hasNfcPassport
        )
    }

    fun mapToIdCountry(map: Map<String, Any?>): IdCountry {
        return IdCountry(
            name = map["name"] as? String ?: "",
            alpha2 = map["alpha2"] as? String ?: "",
            alpha3 = map["alpha3"] as? String ?: "",
            countryCallingCode = map["countryCallingCode"] as? String ?: "",
            hasBarcodeId = map["hasBarcodeId"] as? Boolean ?: false,
            flag = map["flag"] as? String,
            isBanned = map["isBanned"] as? Boolean ?: false,
            hasNfcPassport = map["hasNfcPassport"] as? Boolean ?: false
        )
    }

    // MARK: - AdditionalDataField

    fun mapToAdditionalDataFields(map: Map<String, String>): HashMap<AdditionalDataField, String> {
        val result = HashMap<AdditionalDataField, String>()
        for ((key, value) in map) {
            try {
                // Convert snake_case to SCREAMING_SNAKE_CASE for enum valueOf
                val enumName = key.uppercase()
                val field = AdditionalDataField.valueOf(enumName)
                result[field] = value
            } catch (e: IllegalArgumentException) {
                // Skip unknown fields
            }
        }
        return result
    }

    // MARK: - BusinessConfiguration

    fun businessConfigurationToMap(config: BusinessConfiguration): Map<String, Any?> {
        return mapOf(
            "name" to config.name,
            "identificationId" to config.identificationId,
            "welcomeMessage" to config.welcomeMessage,
            "logoLight" to config.logoLight,
            "minimumAppVersion" to config.minimumAppVersion,
            "phonePinTimeout" to config.phonePinTimeout,
            "phoneRetryAttemptLimit" to config.phoneRetryAttemptLimit,
            "phoneUseAutomaticBypass" to config.phoneUseAutomaticBypass,
            "clientGuid" to config.clientGuid,
            "redirectBackLink" to config.redirectBackLink,
            "completedMessage" to config.completedMessage,
            "isGeolocationEnabled" to config.isGeolocationEnabled
        )
    }

    // MARK: - Picture

    fun pictureToMap(picture: Picture): Map<String, Any?> {
        return mapOf(
            "path" to picture.path,
            "type" to picture.type?.type,
            "barcodes" to picture.barcodes?.map { barcodeToMap(it) },
            "passportContent" to picture.passportContent?.let { passportContentToMap(it) },
            "ePassportContent" to picture.ePassportContent?.let { ePassportContentToMap(it) },
            "livenessInformation" to picture.livenessInformation?.let { livenessInformationToMap(it) }
        )
    }

    // MARK: - Barcode

    private fun barcodeToMap(barcode: com.netki.model.Barcode): Map<String, Any?> {
        return mapOf(
            "rawValue" to barcode.rawValue,
            "displayValue" to barcode.displayValue,
            "valueFormat" to barcode.valueFormat,
            "driverLicense" to barcode.driverLicense?.let { driverLicenseToMap(it) }
        )
    }

    private fun driverLicenseToMap(dl: com.netki.model.Barcode.DriverLicense): Map<String, Any?> {
        return mapOf(
            "documentType" to dl.documentType,
            "documentDiscriminator" to dl.documentDiscriminator,
            "firstName" to dl.firstName,
            "middleName" to dl.middleName,
            "lastName" to dl.lastName,
            "gender" to dl.gender,
            "placeOfBirth" to dl.placeOfBirth,
            "addressStreet" to dl.addressStreet,
            "addressUnit" to dl.addressUnit,
            "addressCity" to dl.addressCity,
            "addressState" to dl.addressState,
            "addressZip" to dl.addressZip,
            "licenseNumber" to dl.licenseNumber,
            "issueDate" to dl.issueDate,
            "expiryDate" to dl.expiryDate,
            "birthDate" to dl.birthDate,
            "height" to dl.height,
            "eyeColor" to dl.eyeColor,
            "hairColor" to dl.hairColor,
            "weight" to dl.weight,
            "issuingCountry" to dl.issuingCountry
        )
    }

    fun picturesToList(pictures: List<Picture>): List<Map<String, Any?>> {
        return pictures.map { pictureToMap(it) }
    }

    // MARK: - PassportContent

    private fun passportContentToMap(content: com.netki.model.PassportContent): Map<String, Any?> {
        return mapOf(
            "text" to content.text,
            "passportType" to content.passportType,
            "countryCode" to content.countryCode,
            "lastName" to content.lastName,
            "firstName" to content.firstName,
            "passportNumber" to content.passportNumber,
            "passportNumberDigitCheck" to content.passportNumberDigitCheck,
            "nationality" to content.nationality,
            "birthDate" to content.birthDate,
            "birthDateDigitCheck" to content.birthDateDigitCheck,
            "gender" to content.gender,
            "expirationDate" to content.expirationDate,
            "expirationDateDigitCheck" to content.expirationDateDigitCheck,
            "personalNumber" to content.personalNumber,
            "personalNumberDigitCheck" to content.personalNumberDigitCheck,
            "stringDigitCheck" to content.stringDigitCheck,
            "digitCheck" to content.digitCheck,
            "isValid" to content.isValid,
            "confidence" to content.confidence
        )
    }

    // MARK: - EPassportContent

    private fun ePassportContentToMap(content: com.netki.model.EPassportContent): Map<String, Any?> {
        return mapOf(
            "isValid" to content.isValid,
            "documentCode" to content.documentCode,
            "issuingState" to content.issuingState,
            "primaryIdentifier" to content.primaryIdentifier,
            "secondaryIdentifier" to content.secondaryIdentifier,
            "nationality" to content.nationality,
            "documentNumber" to content.documentNumber,
            "dateOfBirth" to content.dateOfBirth,
            "gender" to content.gender?.name,
            "dateOfExpiry" to content.dateOfExpiry,
            "optionalData1" to content.optionalData1,
            "optionalData2" to content.optionalData2,
            "faceImagePath" to content.faceImagePath,
            "additionalPersonalData" to content.additionalPersonalData?.let { additionalPersonalDataToMap(it) }
        )
    }

    // MARK: - AdditionalPersonalData

    private fun additionalPersonalDataToMap(data: com.netki.model.AdditionalPersonalData): Map<String, Any?> {
        return mapOf(
            "nameOfHolder" to data.nameOfHolder,
            "custodyInformation" to data.custodyInformation,
            "otherValidTDNumbers" to data.otherValidTDNumbers,
            "permanentAddress" to data.permanentAddress,
            "personalNumber" to data.personalNumber,
            "placeOfBirth" to data.placeOfBirth,
            "telephone" to data.telephone,
            "profession" to data.profession,
            "title" to data.title,
            "personalSummary" to data.personalSummary,
            "proofOfCitizenship" to data.proofOfCitizenship,
            "tdNumbers" to data.tdNumbers
        )
    }

    // MARK: - LivenessInformation

    private fun livenessInformationToMap(info: com.netki.model.LivenessInformation): Map<String, Any?> {
        return mapOf(
            "isLive" to info.isLive,
            "livenessScore" to info.livenessScore,
            "livenessActionAttempts" to info.livenessActionAttempts
        )
    }

    // MARK: - OnBoardIdTheme

    fun mapToOnBoardIdTheme(map: Map<String, Any?>): OnBoardIdTheme {
        @Suppress("UNCHECKED_CAST")
        val buttonsMap = map["buttons"] as? Map<String, Any?> ?: emptyMap()
        @Suppress("UNCHECKED_CAST")
        val contentMap = map["content"] as? Map<String, Any?> ?: emptyMap()
        @Suppress("UNCHECKED_CAST")
        val instructionsList = map["instructions"] as? List<Map<String, Any?>> ?: emptyList()

        return OnBoardIdTheme(
            buttons = mapToButtonColors(buttonsMap),
            content = mapToContentColors(contentMap),
            instructions = instructionsList.map { mapToInstructionContent(it) }
        )
    }

    private fun mapToButtonColors(map: Map<String, Any?>): OnBoardIdTheme.ButtonColors {
        return OnBoardIdTheme.ButtonColors(
            primaryBackground = map["primaryBackground"] as? String,
            primaryText = map["primaryText"] as? String,
            secondaryBackground = map["secondaryBackground"] as? String,
            secondaryText = map["secondaryText"] as? String,
            cornerRadiusDp = (map["cornerRadiusDp"] as? Number)?.toFloat()
        )
    }

    private fun mapToContentColors(map: Map<String, Any?>): OnBoardIdTheme.ContentColors {
        return OnBoardIdTheme.ContentColors(
            background = map["background"] as? String,
            surface = map["surface"] as? String,
            border = map["border"] as? String,
            titleText = map["titleText"] as? String,
            subtitleText = map["subtitleText"] as? String,
            bodyText = map["bodyText"] as? String
        )
    }

    private fun mapToInstructionContent(map: Map<String, Any?>): OnBoardIdTheme.InstructionContent {
        val idTypeStr = map["idType"] as? String ?: ""
        val pictureTypeStr = map["pictureType"] as? String ?: ""

        @Suppress("UNCHECKED_CAST")
        val stepsList = map["steps"] as? List<Map<String, Any?>>

        return OnBoardIdTheme.InstructionContent(
            idType = IdType.entries.firstOrNull { it.type == idTypeStr } ?: IdType.GOVERNMENT_ID,
            pictureType = PictureType.entries.firstOrNull { it.type == pictureTypeStr } ?: PictureType.FRONT,
            title = map["title"] as? String,
            subtitle = map["subtitle"] as? String,
            body = map["body"] as? String,
            steps = stepsList?.map { mapToInstructionItem(it) },
            imageUrl = map["imageUrl"] as? String,
            videoPath = map["videoPath"] as? String
        )
    }

    private fun mapToInstructionItem(map: Map<String, Any?>): OnBoardIdTheme.InstructionItem {
        return OnBoardIdTheme.InstructionItem(
            text = map["text"] as? String ?: "",
            iconUrl = map["iconUrl"] as? String,
            isWarning = map["isWarning"] as? Boolean ?: false
        )
    }
}

File: ios/Classes/Mappers.swift
import NetkiSDK

struct Mappers {

    // MARK: - ResultInfo

    static func resultInfoToMap(_ resultInfo: ResultInfo) -> [String: Any?] {
        return [
            "status": resultInfo.status == .SUCCESS ? "SUCCESS" : "ERROR",
            "extraData": sanitizeForFlutter(resultInfo.extraData),
            "errorType": resultInfo.errorType?.rawValue,
            "message": resultInfo.message
        ]
    }

    // MARK: - Sanitize for Flutter

    /// Recursively sanitize data to ensure all values are Flutter-compatible
    static func sanitizeForFlutter(_ value: Any?) -> Any? {
        guard let value = value else { return nil }

        switch value {
        case let string as String:
            return string
        case let number as NSNumber:
            return number
        case let bool as Bool:
            return bool
        case let array as [Any]:
            return array.map { sanitizeForFlutter($0) }
        case let dict as [String: Any]:
            return dict.mapValues { sanitizeForFlutter($0) }
        case let picture as Picture:
            return pictureToMap(picture)
        default:
            // For any other type, convert to string representation
            return String(describing: value)
        }
    }

    // MARK: - Picture

    static func pictureToMap(_ picture: Picture) -> [String: Any?] {
        return [
            "path": picture.path,
            "type": picture.type.rawValue,
            "barcodes": picture.barcodes?.map { barcodeToMap($0) },
            "passportContent": picture.passportContent.map { passportContentToMap($0) },
            "ePassportContent": picture.ePassportContent.map { ePassportContentToMap($0) },
            "livenessInformation": picture.livenessInformation.map { livenessInformationToMap($0) }
        ]
    }

    private static func barcodeToMap(_ barcode: Barcode) -> [String: Any?] {
        return [
            "rawValue": barcode.rawValue,
            "displayValue": barcode.displayValue,
            "valueFormat": barcode.valueFormat,
            "driverLicense": barcode.driverLicense.map { driverLicenseToMap($0) }
        ]
    }

    private static func driverLicenseToMap(_ dl: DriverLicense) -> [String: Any?] {
        return [
            "documentType": dl.documentType,
            "documentDiscriminator": dl.documentDiscriminator,
            "firstName": dl.firstName,
            "middleName": dl.middleName,
            "lastName": dl.lastName,
            "gender": dl.gender,
            "placeOfBirth": dl.placeOfBirth,
            "addressStreet": dl.addressStreet,
            "addressUnit": dl.addressUnit,
            "addressCity": dl.addressCity,
            "addressState": dl.addressState,
            "addressZip": dl.addressZip,
            "licenseNumber": dl.licenseNumber,
            "issueDate": dl.issueDate,
            "expiryDate": dl.expiryDate,
            "birthDate": dl.birthDate,
            "height": dl.height,
            "eyeColor": dl.eyeColor,
            "hairColor": dl.hairColor,
            "weight": dl.weight,
            "issuingCountry": dl.issuingCountry
        ]
    }

    private static func passportContentToMap(_ content: PassportContent) -> [String: Any?] {
        return [
            "text": content.text,
            "passportType": content.passportType,
            "countryCode": content.countryCode,
            "lastName": content.lastName,
            "firstName": content.firstName,
            "passportNumber": content.passportNumber,
            "passportNumberDigitCheck": content.passportNumberDigitCheck,
            "nationality": content.nationality,
            "birthDate": content.birthDate,
            "birthDateDigitCheck": content.birthDateDigitCheck,
            "gender": content.gender,
            "expirationDate": content.expirationDate,
            "expirationDateDigitCheck": content.expirationDateDigitCheck,
            "personalNumber": content.personalNumber,
            "personalNumberDigitCheck": content.personalNumberDigitCheck,
            "stringDigitCheck": content.stringDigitCheck,
            "digitCheck": content.digitCheck,
            "isValid": content.isValid,
            "confidence": content.confidence
        ]
    }

    private static func ePassportContentToMap(_ content: EPassportContent) -> [String: Any?] {
        return [
            "isValid": content.isValid,
            "documentCode": content.documentCode,
            "issuingState": content.issuingState,
            "primaryIdentifier": content.primaryIdentifier,
            "secondaryIdentifier": content.secondaryIdentifier,
            "nationality": content.nationality,
            "documentNumber": content.documentNumber,
            "dateOfBirth": content.dateOfBirth,
            "gender": content.gender?.rawValue,
            "dateOfExpiry": content.dateOfExpiry,
            "optionalData1": content.optionalData1,
            "optionalData2": content.optionalData2,
            "faceImagePath": content.faceImagePath,
            "additionalPersonalData": content.additionalPersonalData.map { additionalPersonalDataToMap($0) }
        ]
    }

    private static func additionalPersonalDataToMap(_ data: AdditionalPersonalData) -> [String: Any?] {
        return [
            "nameOfHolder": data.nameOfHolder,
            "custodyInformation": data.custodyInformation,
            "otherValidTDNumbers": data.otherValidTDNumbers,
            "permanentAddress": data.permanentAddress,
            "personalNumber": data.personalNumber,
            "placeOfBirth": data.placeOfBirth,
            "telephone": data.telephone,
            "profession": data.profession,
            "title": data.title,
            "personalSummary": data.personalSummary,
            "proofOfCitizenship": data.proofOfCitizenship,
            "tdNumbers": data.tdNumbers
        ]
    }

    private static func livenessInformationToMap(_ info: LivenessInformation) -> [String: Any?] {
        return [
            "isLive": info.isLive,
            "livenessScore": info.livenessScore,
            "livenessActionAttempts": info.livenessActionAttempts
        ]
    }

    // MARK: - IdCountry

    static func idCountryToMap(_ country: IdCountry) -> [String: Any?] {
        return [
            "name": country.name,
            "alpha2": country.alpha2,
            "alpha3": country.alpha3,
            "countryCallingCode": country.countryCallingCode,
            "hasBarcodeId": country.hasBarcodeId,
            "flag": country.flag,
            "isBanned": country.isBanned,
            "hasNfcPassport": country.hasNfcPassport
        ]
    }

    static func mapToIdCountry(_ map: [String: Any]) -> IdCountry {
        return IdCountry(
            name: map["name"] as? String ?? "",
            alpha2: map["alpha2"] as? String ?? "",
            alpha3: map["alpha3"] as? String ?? "",
            countryCallingCode: map["countryCallingCode"] as? String ?? "",
            hasBarcodeId: map["hasBarcodeId"] as? Bool ?? false,
            flag: map["flag"] as? String,
            isBanned: map["isBanned"] as? Bool ?? false,
            hasNfcPassport: map["hasNfcPassport"] as? Bool ?? false
        )
    }

    // MARK: - AdditionalDataField

    static func mapToAdditionalDataFields(_ map: [String: String]) -> [AdditionalDataField: String] {
        var result: [AdditionalDataField: String] = [:]
        for (key, value) in map {
            if let field = AdditionalDataField(rawValue: key) {
                result[field] = value
            }
        }
        return result
    }

    // MARK: - BusinessConfiguration

    static func businessConfigurationToMap(_ config: BusinessConfiguration) -> [String: Any?] {
        return [
            "name": config.name,
            "identificationId": config.identificationId,
            "welcomeMessage": config.welcomeMessage,
            "logoLight": config.logoLight,
            "minimumAppVersion": config.minimumAppVersion,
            "phonePinTimeout": config.phonePinTimeout,
            "phoneRetryAttemptLimit": config.phoneRetryAttemptLimit,
            "phoneUseAutomaticBypass": config.phoneUseAutomaticBypass,
            "clientGuid": config.clientGuid,
            "redirectBackLink": config.redirectBackLink,
            "completedMessage": config.completedMessage,
            "isGeolocationEnabled": config.isGeolocationEnabled
        ]
    }

    // MARK: - OnBoardIdTheme

    static func mapToOnBoardIdTheme(_ map: [String: Any]) -> OnBoardIdTheme {
        let buttonsMap = map["buttons"] as? [String: Any] ?? [:]
        let contentMap = map["content"] as? [String: Any] ?? [:]
        let instructionsList = map["instructions"] as? [[String: Any]] ?? []

        return OnBoardIdTheme(
            buttons: mapToButtonColors(buttonsMap),
            content: mapToContentColors(contentMap),
            instructions: instructionsList.map { mapToInstructionContent($0) }
        )
    }

    private static func mapToButtonColors(_ map: [String: Any]) -> OnBoardIdTheme.ButtonColors {
        return OnBoardIdTheme.ButtonColors(
            primaryBackground: map["primaryBackground"] as? String,
            primaryText: map["primaryText"] as? String,
            secondaryBackground: map["secondaryBackground"] as? String,
            secondaryText: map["secondaryText"] as? String,
            cornerRadiusDp: (map["cornerRadiusDp"] as? NSNumber).map { CGFloat($0.doubleValue) }
        )
    }

    private static func mapToContentColors(_ map: [String: Any]) -> OnBoardIdTheme.ContentColors {
        return OnBoardIdTheme.ContentColors(
            background: map["background"] as? String,
            surface: map["surface"] as? String,
            border: map["border"] as? String,
            titleText: map["titleText"] as? String,
            subtitleText: map["subtitleText"] as? String,
            bodyText: map["bodyText"] as? String
        )
    }

    private static func mapToInstructionContent(_ map: [String: Any]) -> OnBoardIdTheme.InstructionContent {
        let idTypeStr = map["idType"] as? String ?? ""
        let pictureTypeStr = map["pictureType"] as? String ?? ""
        let stepsList = map["steps"] as? [[String: Any]]

        return OnBoardIdTheme.InstructionContent(
            idType: IdType(rawValue: idTypeStr) ?? .GOVERNMENT_ID,
            pictureType: PictureType(rawValue: pictureTypeStr) ?? .FRONT,
            title: map["title"] as? String,
            subtitle: map["subtitle"] as? String,
            body: map["body"] as? String,
            steps: stepsList?.map { mapToInstructionItem($0) },
            imageUrl: map["imageUrl"] as? String,
            videoPath: map["videoPath"] as? String
        )
    }

    private static func mapToInstructionItem(_ map: [String: Any]) -> OnBoardIdTheme.InstructionItem {
        return OnBoardIdTheme.InstructionItem(
            text: map["text"] as? String ?? "",
            iconUrl: map["iconUrl"] as? String,
            isWarning: map["isWarning"] as? Bool ?? false
        )
    }
}
